bundles/scripting/core/src/main/java/org/apache/sling/scripting/core/impl/DefaultSlingScript.java

```diff
@@ -371,21 +377,31 @@ public void setReader(Reader reader) {
             }
 
             // evaluate the script
-            final Object result;
-            if (method == null && this.scriptEngine instanceof Compilable) {
-                CachedScript cachedScript = scriptCache.getScript(scriptName);
-                if (cachedScript == null) {
-                    ScriptNameAwareReader snReader = new ScriptNameAwareReader(reader, scriptName);
-                    CompiledScript compiledScript = ((Compilable) scriptEngine).compile(snReader);
-                    cachedScript = new CachedScriptImpl(scriptName, compiledScript);
-                    scriptCache.putScript(cachedScript);
-                    LOGGER.debug("Adding {} to the script cache.", scriptName);
+            Object result = null;
+            String selectorString = props.getRequest().getRequestPathInfo().getSelectorString();
+            if(StringUtils.isBlank(selectorString) || !selectorString.contains(BVP_VARIABLES_SELECTOR)) {
+                if (method == null && this.scriptEngine instanceof Compilable) {
+                    CachedScript cachedScript = scriptCache.getScript(scriptName);
+                    if (cachedScript == null) {
+                        ScriptNameAwareReader snReader = new ScriptNameAwareReader(reader, scriptName);
+                        CompiledScript compiledScript = ((Compilable) scriptEngine).compile(snReader);
+                        cachedScript = new CachedScriptImpl(scriptName, compiledScript);
+                        scriptCache.putScript(cachedScript);
+                        LOGGER.debug("Adding {} to the script cache.", scriptName);
+                    } else {
+                        LOGGER.debug("Script {} was already cached.", scriptName);
+                    }
+                    result = cachedScript.getCompiledScript().eval(ctx);
                 } else {
-                    LOGGER.debug("Script {} was already cached.", scriptName);
+                    result = scriptEngine.eval(reader, ctx);
                 }
-                result = cachedScript.getCompiledScript().eval(ctx);
             } else {
-                result = scriptEngine.eval(reader, ctx);
+                acquireScriptingVariables(ctx, bvpVariables, 100, scriptEngine.getFactory().getEngineName());
+                acquireScriptingVariables(ctx, bvpVariables, 200, "GLOBAL");
+                acquireScriptingVariables(ctx, bvpVariables, -314, "SLING");
+                Gson  gson = new Gson();
```

Comment:
Use commons.json to construct the json manually within the ScriptingVariableModel. We should be very careful before introducting additional 3rd party libraries.